### PR TITLE
lib/dotinstall: add file concatenation function

### PR DESCRIPTION
**Pull request by mard**
# Change description
Adds a function that concatenates multiple source files into one.

# Use case
Allows modularization of configuration for programs that do not allow sourcing multiple or non-existent files.

For example, an application expects a single `.config/i3/config` file, but you need multiple configurations for different setups. In case of `i3` configuration, there is no known way to include a file that might not exist, so the only choice is concatenation.

Optionally a comment format can be specified to include some verbose information with concatenation details.

# Usage

```python
util.concat_files(['./config1', '/path/to/config2'],
                  '~/.config/myconfig',
                  '# %s')`
```
# Example
## File tree
```
📁/cfg/test
├──📄 config-base
├──📄 config-work
└──📁 subdir-test1
    └──📁 subdir-test2
        └──📄 config-home
```

## Install `/cfg/test/install.py`
```python
import socket
from dotinstall import util

def run():
    sources = ['./config-base']
    if socket.gethostname() == 'MyWorkMachine':
        sources += ['./config-work']
    else:
        sources += ['./subdir-test1/subdir-test2/config-home']
    util.concat_files(sources, '~/.config/test/config', '# %s')
```

## Output `~/.config/test/config`
```gitconfig
# DO NOT EDIT THIS FILE
# This file was concatenated by dotfiles utility.
# All changes in this file will not be retained.

# Concatenated file 1 of 2 from: /home/mard/repos/ext/dotfiles-rr/cfg/test/config-base

contents of config-base...

# Concatenated file 2 of 2 from: /home/mard/repos/ext/dotfiles-rr/cfg/test/subdir-test1/subdir-test2/config-home

contents of config-home...
```

# Development configuration

* Arch Linux 5.8.3-arch1-1
* Python 3.8.5
* pylint 2.5.3

### CODE REVIEW

File: lib/dotinstall/util.py [hunk 1]
```diff
@@ -115,6 +115,31 @@ def create_symlink(source: str, target: str) -> None:
     os.symlink(source, target)
 
 
+def concat_files(sources, target, comment_prefix=None):
+    for i, _ in enumerate(sources):
+        sources[i] = os.path.abspath(os.path.expanduser(sources[i]))
```
Comment (rr-) [inline lib/dotinstall/util.py:120]:
```python3
sources = [
    os.path.abspath(os.path.expanduser(path))
    for path in sources
]
```

File: lib/dotinstall/util.py [hunk 2]
```diff
@@ -115,6 +115,31 @@ def create_symlink(source: str, target: str) -> None:
...
+    target = os.path.expanduser(target)
+    _remove_symlink(target)
+    create_dir(os.path.dirname(target))
+    print('Concatenating %s to %s...' % (sources, target))
```
Comment (rr-) [inline lib/dotinstall/util.py:124]:
```diff
-    print('Concatenating %s to %s...' % (sources, target))
+    logger.info('Concatenating %s to %s...', sources, target)
```

File: lib/dotinstall/util.py [hunk 3]
```diff
@@ -115,6 +115,31 @@ def create_symlink(source: str, target: str) -> None:
...
+    with open(target, 'w') as ft:
```
Comment (rr-) [inline lib/dotinstall/util.py:125]: Please pass the file through [black](https://github.com/psf/black)

```
python3 -m pip install --user black
black -l79 path/to/your/fork
```

File: lib/dotinstall/util.py [hunk 4]
```diff
@@ -115,6 +115,31 @@ def create_symlink(source: str, target: str) -> None:
     os.symlink(source, target)
 
 
+def concat_files(sources, target, comment_prefix=None):
```
Comment (rr-) [inline lib/dotinstall/util.py:118]: Let's make `comment_prefix` into `comment_fmt`

File: lib/dotinstall/util.py [hunk 5]
```diff
@@ -115,6 +115,31 @@ def create_symlink(source: str, target: str) -> None:
...
+    create_dir(os.path.dirname(target))
+    print('Concatenating %s to %s...' % (sources, target))
+    with open(target, 'w') as ft:
+        if comment_prefix:
+            ft.write('%sDO NOT EDIT THIS FILE%s'
+                     % (comment_prefix, os.linesep))
+            ft.write('%sThis file was concatenated by dotfiles utility.%s'
+                     % (comment_prefix, os.linesep))
+            ft.write('%sAll changes in this file will not be retained.%s%s'
+                     % (comment_prefix, os.linesep, os.linesep))
```
Comment (rr-) [inline lib/dotinstall/util.py:132]: …and here use
```python3
print(comment_fmt.format('your message…'), file=handle)
```
that way you don't need os.linesep, `print`  will take care of it for you.
To print a blank line to a file, no param is needed:
```python3
print(file=handle)
```

File: lib/dotinstall/util.py (hunk 3 above)
Comment (rr-) [reply lib/dotinstall/util.py:125]: Let's name `ft` a `handle`

File: lib/dotinstall/util.py [hunk 6]
```diff
@@ -115,6 +115,31 @@ def create_symlink(source: str, target: str) -> None:
...
+        for source in sources:
+            with open(source) as fs:
+                if comment_prefix:
+                    ft.write('%sConcatenated file %s of %s from: %s%s%s'
+                             % (comment_prefix, sources.index(source)+1,
+                                len(sources), source, os.linesep, os.linesep))
```
Comment (rr-) [inline lib/dotinstall/util.py:138]: Similar remark as above

File: lib/dotinstall/util.py [hunk 7]
```diff
@@ -115,6 +115,31 @@ def create_symlink(source: str, target: str) -> None:
...
+                ft.write(fs.read())
+                ft.write(os.linesep)
```
Comment (rr-) [inline lib/dotinstall/util.py:140]: Let's change the logic a little bit and make sure the concatenated files do not end with a double line separator (e.g. one blank line):

```python3
for i, source in enumerate(sources):
    if i > 0:
        print(file=handle)
    print(fs.read(), file=handle)
```

also to ensure better scope let's make sure the source path is opened only for the read operation, not for the whole write block

**Comment by rr-** [comment]
Ideally if you feel like it you could use `pathlib.Path` here; that way you can avoid context managers altogether and use `Path(…).read_text()` to read the entire file at once.

File: lib/dotinstall/util.py (hunk 4 above)
Comment (rr-) [reply lib/dotinstall/util.py:118]: Let's add type hints: `sources: T.List[str]`, `target: str`, `comment_prefix: T.Optional[str]`, `-> None`

**Comment by mard** [comment]
Last parameter is now a format string.

To specify a comment format you can do so like:

```python
util.concat_files(['./config1', '/path/to/config2'], '~/.config/myconfig', '# %s')`
```

File: lib/dotinstall/util.py (hunk 1 above)
Comment (mard) [reply lib/dotinstall/util.py]: Done.

File: lib/dotinstall/util.py (hunk 2 above)
Comment (mard) [reply lib/dotinstall/util.py]: Done.

File: lib/dotinstall/util.py (hunk 3 above)
Comment (mard) [reply lib/dotinstall/util.py]: Done.

File: lib/dotinstall/util.py (hunk 4 above)
Comment (mard) [reply lib/dotinstall/util.py]: Done.

File: lib/dotinstall/util.py (hunk 5 above)
Comment (mard) [reply lib/dotinstall/util.py]: Done.

File: lib/dotinstall/util.py (hunk 6 above)
Comment (mard) [reply lib/dotinstall/util.py]: Done.

File: lib/dotinstall/util.py (hunk 7 above)
Comment (mard) [reply lib/dotinstall/util.py]: Opted for `pathlib.Path` approach. Resolving.

Comment (rr-) [reply lib/dotinstall/util.py]: The first part of this remark is still to be resolved

File: lib/dotinstall/util.py [hunk 8]
```diff
@@ -115,6 +115,39 @@ def create_symlink(source: str, target: str) -> None:
...
+                comment_fmt
+                % "This file was concatenated by dotfiles utility.",
+                file=target_handle,
+            )
+            print(
+                comment_fmt % "All changes in this file will not be retained.",
+                file=target_handle,
+            )
+            print(file=target_handle)
+        for source in sources:
```
Comment (rr-) [inline lib/dotinstall/util.py:139]: I'd put a blank line before that `for` loop

File: lib/dotinstall/util.py [hunk 9]
```diff
@@ -115,6 +115,39 @@ def create_symlink(source: str, target: str) -> None:
...
 
 
+def concat_files(
+    sources: T.List[str], target: str, comment_fmt: T.Optional[str] = None
+) -> None:
+    sources = [os.path.abspath(os.path.expanduser(path)) for path in sources]
+    target = os.path.expanduser(target)
+    _remove_symlink(target)
+    create_dir(os.path.dirname(target))
+    logger.info("Concatenating %s to %s...", sources, target)
```
Comment (rr-) [inline lib/dotinstall/util.py:125]: And before `logger.info` as well

File: lib/dotinstall/util.py (hunk 8 above)
Comment (mard) [reply lib/dotinstall/util.py:141]: Not valid with Black
https://github.com/rr-/dotfiles/pull/13/checks?check_run_id=1215213028

Comment (mard) [reply lib/dotinstall/util.py:141]: Oh, nevermind. Those were empty spaces!

File: lib/dotinstall/util.py (hunk 7 above)
Comment (mard) [reply lib/dotinstall/util.py]: Oh right, done.

**Review by rr- — APPROVED** [review]

**Pull request merged**